code/parachain/runtime/dali/src/lib.rs

```diff
@@ -1196,6 +1211,157 @@ impl cosmwasm::Config for Runtime {
+		match module_id.borrow().to_string().as_str() {
+			pallet_ibc_ping::MODULE_ID => Some(&mut self.pallet_ibc_ping),
+			_ => None,
+		}
+	}
+
+	fn has_route(module_id: &impl Borrow<ModuleId>) -> bool {
+		matches!(module_id.borrow().to_string().as_str(), pallet_ibc_ping::MODULE_ID)
+	}
+
+	fn lookup_module_by_port(port_id: &PortId) -> Option<ModuleId> {
+		match port_id.as_str() {
+			pallet_ibc_ping::PORT_ID => ModuleId::from_str(pallet_ibc_ping::MODULE_ID).ok(),
+			_ => None,
+		}
+	}
+}
+
+pub struct IbcDenomToAssetIdConversion;
+
+impl DenomToAssetId<Runtime> for IbcDenomToAssetIdConversion {
+	type Error = DispatchError;
+
+	fn from_denom_to_asset_id(denom: &String) -> Result<CurrencyId, Self::Error> {
+		let denom_bytes = denom.as_bytes().to_vec();
+		if let Some(id) = IbcDenoms::<Runtime>::get(&denom_bytes) {
+			return Ok(id)
+		}
+
+		// let pallet_id: AccountId = PalletId(*b"pall-ibc").into_account_truncating();
+
+		// let symbol = denom
+		// 	.split("/")
+		// 	.last()
+		// 	.ok_or_else(|| DispatchError::Other("denom missing a name"))?
+		// 	.as_bytes()
+		// 	.to_vec();
+		// let asset_id = generate_asset_id()?;
+
+		// <assets::Pallet<Runtime> as Create<AccountId>>::create(
+		// 	asset_id,
+		// 	pallet_id.clone(),
+		// 	true,
+		// 	1,
+		// )?;
+		//
+		// <assets::Pallet<Runtime> as Mutate<AccountId>>::set(
+		// 	asset_id,
+		// 	&pallet_id,
+		// 	denom_bytes,
+		// 	symbol,
+		// 	12,
+		// )?;
+		// TODO: the code above is outdated
+		let asset_id =
+			<currency_factory::Pallet<Runtime> as CurrencyFactoryT>::create(RangeId::IBC_ASSETS)?;
+
+		IbcDenoms::<Runtime>::insert(denom_bytes.clone(), asset_id);
+		IbcAssetIds::<Runtime>::insert(asset_id, denom_bytes.clone());
+
+		let location = XcmAssetLocation::new(
+			MultiLocation::new(1, X1(Junction::GeneralIndex(asset_id.into()))).into(),
+		);
+		assets_registry::Pallet::<Runtime>::set_reserve_location(
+			asset_id,
+			location,
+			Rational64::one(),
+			Some(12),
+		)?;
+
+		Ok(asset_id)
+	}
+
+	fn from_asset_id_to_denom(id: CurrencyId) -> Option<String> {
+		// let name = <assets::Pallet<Runtime> as InspectMetadata<AccountId>>::name(&id);
+		// String::from_utf8(name).ok()
+		// TODO: the code above is outdated
+		IbcAssetIds::<Runtime>::get(id)
+			.map(|denom| String::from_utf8(denom).ok())
+			.flatten()
+	}
+
+	fn ibc_assets(
+		start_key: Option<CurrencyId>,
+		offset: Option<u32>,
+		mut limit: u64,
+	) -> (Vec<Vec<u8>>, u64, Option<CurrencyId>) {
+		let mut iterator = if let Some(asset_id) = start_key {
+			let raw_key = asset_id.encode();
+			IbcAssetIds::<Runtime>::iter_from(raw_key).skip(0)
+		} else if let Some(offset) = offset {
+			IbcAssetIds::<Runtime>::iter().skip(offset as usize)
+		} else {
+			IbcAssetIds::<Runtime>::iter().skip(0)
+		};
+
+		let mut denoms = vec![];
+		for (_, denom) in iterator.by_ref() {
+			denoms.push(denom);
+			limit -= 1;
+			if limit == 0 {
+				break
+			}
+		}
```

Comment:
Can be something like

```
iterator.by_ref().take(limit).collect().
```

Technically you are also returning both an iterator and denoms, which contain the same value? Bit of a codesmell but don't think it needs to be fixed in this PR.

Reply:
Thanks. No, the iterator is not returned, there is a call to `.next`